Add linux/io_uring.h and linux/fs.h bindings

ljmf00 opened the pull request:
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

---

`io_uring` bindings support are extremely important for implementing efficient IO for future eventloop drivers coming to D programming language.

These headers were generated and manually adapted. `linux/io_uring.h` is dependent of `linux/fs.h` due to `__kernel_rwf_t` type and its `RWF_*` flags.

I'm not entirely sure what's the interface of linux headers on the runtime, but I think those headers have a good fit here.

ibuclaw commented on src/core/sys/linux/fs.d:
```diff
@@ -0,0 +1,274 @@
+
+/* See blkpg.h */
+
+/* Some people are morons.  Do not use sizeof! */
```
Do we really need all comments included ad verbatim? Just ddoc `///` markers will be enough.

ljmf00 replied:
Yeah, I'm going to remove some of those from here and transform what should be useful for documentation to Ddoc comments.

ljmf00 replied:
Done

RazvanN7 commented:
What's the status of this draft? Is it ready for review?

RazvanN7 commented:
ping @ljmf00

ljmf00 commented:
> What's the status of this draft? Is it ready for review?

> ping @ljmf00

I'm sorry for leaving this hanging. I had my brain "damaged" for a while and decided to take a little break to hold some ideas and prepare for my final course semester. I'm goingo to fix it today and make it ready for review, as soon as possible.

ljmf00 commented:
@RazvanN7 I forgot to push :disappointed: . Although, you can ping me whenever a PR made by me is stalled. I have now the time and brain capacity to deal with D contributions. I changed the documentation style to conform with the D style and rebased it with the latest commit. I'm so sorry for this inconvenience.

ljmf00 commented:
Picked the wrong version

Geod24 reviewed (commented):
Would be great to have, ping me when ready for review.

ljmf00 commented on src/core/sys/linux/fs.d:
```diff
@@ -0,0 +1,274 @@
+@system:
+@nogc:
+nothrow:
+@system:
```
Done

ljmf00 commented:
@Geod24 ping as requested. Sorry for my messy discussion on this PR. It may end up being confusing with all these changes. I rebased and edited the commit messages to match implementation and removed useless comments from upstream. Please check if this is good to go now.

RazvanN7 commented:
Looks good to me. @Geod24 merge when you think it's ready.

Geod24 commented on src/core/sys/linux/io_uring.d:
```diff
@@ -0,0 +1,414 @@
+/**
+ * D header file for the io_uring interface.
+ * Available since Linux 5.1
```
Newline to separate the title from the description. Also some of those symbols are only available since 5.5.

Geod24 commented on src/core/sys/linux/io_uring.d:
```diff
@@ -0,0 +1,414 @@
+        uint fadvise_advice;
+        uint splice_flags;
+        uint rename_flags;
+        uint unlink_flags;
```
The latest version also have a hard `hardlink_flags`.

Geod24 commented on src/core/sys/linux/io_uring.d:
```diff
@@ -0,0 +1,414 @@
+    IORING_REGISTER_PERSONALITY = 9,
+    IORING_UNREGISTER_PERSONALITY = 10,
+    IORING_REGISTER_RESTRICTIONS = 11,
+    IORING_REGISTER_ENABLE_RINGS = 12,
```
Latest version has some `****2` members.

Geod24 commented on src/core/sys/linux/io_uring.d:
```diff
@@ -0,0 +1,414 @@
+@system:
+@nogc:
+nothrow:
+@system:
```
Duplicated

Geod24 commented on src/core/sys/linux/fs.d:
```diff
@@ -0,0 +1,265 @@
+    ushort dest_count; /// in - total elements in info array
+    ushort reserved1; /// must be zero
+    uint reserved2; /// must be zero
+    file_dedupe_range_info[0] info;
```
Flexible array here too

Geod24 reviewed (commented):
Sorry for the delay! Gave it a review, looks very good. The only things I noted are a couple style nitpicks, some new members (but that might not even be released, it was definitely submitted to the kernel tho) and the usage of flexible array member of which I'm not too sure about.
